### PR TITLE
Base implementation of IDisk interafce for S3

### PR DESCRIPTION
**Pull request by GrigoryPervakov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (up to few sentences, required except for Non-significant/Documentation categories):

Add base implementation for IDisk interface for S3 storage

### CODE REVIEW

File: dbms/tests/integration/test_tinylog_s3/test.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+
+def test_tinylog_s3(cluster):
+    node = cluster.instances["node"]
+    minio = cluster.minio_client
+
+    node.query("CREATE TABLE s3_test (id UInt64) Engine=TinyLog")
+    node.query("INSERT INTO s3_test SELECT number FROM numbers(3)")
+    assert node.query("SELECT * FROM s3_test") == "0\n1\n2\n"
+    assert list(map(lambda obj: obj.object_name[:-17],
```
Comment (Alex-Burmak) [inline dbms/tests/integration/test_tinylog_s3/test.py]: What's a magic number 17?

File: dbms/src/Disks/DiskS3.h [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+    void createDirectory(const String & path) override;
+
+    void createDirectories(const String & path) override;
+
+    void clearDirectory(const String & path) override
+    {
+        for (auto it{iterateDirectory(path)}; it->isValid(); it->next())
+            remove(it->path());
+    }
```
Comment (Alex-Burmak) [inline dbms/src/Disks/DiskS3.h]: The `clearDirectory` method should removes only files (see https://github.com/ClickHouse/ClickHouse/pull/8356#discussion_r362129176). Probably we need to add some option to `clearDirectory` for removing directories as well.

File: dbms/src/Storages/StorageTinyLog.cpp [hunk 3]
```diff
@@ -428,6 +428,12 @@ void StorageTinyLog::truncate(const ASTPtr &, const Context &, TableStructureWri
         addFiles(column.name, *column.type);
 }
 
+void StorageTinyLog::drop(TableStructureWriteLockHolder &)
+{
+    std::unique_lock<std::shared_mutex> lock(rwlock);
+    disk->clearDirectory(table_path);
```
Comment (Alex-Burmak) [inline dbms/src/Storages/StorageTinyLog.cpp]: Shouldn't `table_path` itself be deleted as well?

File: dbms/src/Disks/DiskS3.cpp [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+        from_file.renameTo(to_file.path());
+}
+
+void DiskS3::copyFile(const String & from_path, const String & to_path)
+{
+    if (exists(to_path))
+        remove(to_path);
+
+    String s3_from_path;
+    String s3_to_path = s3_root_path + to_path + getRandomSuffix();
```
Comment (Alex-Burmak) [inline dbms/src/Disks/DiskS3.cpp]: To better decouple data and metadata I'd generate `s3_to_path` something like  `s3_root_path + '.storage/' + getRandomFilename()`.

File: dbms/tests/integration/test_tinylog_s3/test.py (hunk 1 above)
Comment (GrigoryPervakov) [reply dbms/tests/integration/test_tinylog_s3/test.py]: Current s3 object name is root_path + path + '.' + random string of 16 chars. Here I cut random suffix with a dot.

File: dbms/src/Storages/StorageTinyLog.cpp (hunk 3 above)
Comment (GrigoryPervakov) [reply dbms/src/Storages/StorageTinyLog.cpp]: Then we need to add `void remove(bool recursive);` in IDisk interface.

File: dbms/src/Disks/DiskS3.cpp (hunk 4 above)
Comment (GrigoryPervakov) [reply dbms/src/Disks/DiskS3.cpp]: I had preferred to add origin filename to make bucket content more human-readable and reduce the probability of name overlap.

File: dbms/src/Disks/DiskMemory.cpp [hunk 5]
```diff
@@ -231,6 +231,34 @@ std::unique_ptr<WriteBuffer> DiskMemory::writeFile(const String & path, size_t /
         return std::make_unique<WriteBufferFromString>(iter->second.data);
 }
 
+void DiskMemory::remove(const String & path, bool recursive)
+{
+    std::lock_guard lock(mutex);
+
+    auto file_it = files.find(path);
+    if (file_it == files.end())
+        return;
```
Comment (alexey-milovidov) [inline dbms/src/Disks/DiskMemory.cpp]: Why don't throw an exception?

File: dbms/src/Disks/DiskS3.h [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+#    include <aws/s3/S3Client.h>
+#    include <Poco/DirectoryIterator.h>
+
+
+namespace DB
+{
+/**
+ * Storage for persisting data in S3 and metadata on the local disk.
+ * Files are represented by file in local filesystem (clickhouse_root/disks/disk_name/path/to/file)
+ * that contains S3 object key with actual data.
```
Comment (alexey-milovidov) [inline dbms/src/Disks/DiskS3.h]: And also should contain file size.

File: dbms/src/Disks/IDisk.h [hunk 7]
```diff
@@ -125,6 +126,13 @@ class IDisk : public Space
 
     /// Open the file for write and return WriteBuffer object.
     virtual std::unique_ptr<WriteBuffer> writeFile(const String & path, size_t buf_size = DBMS_DEFAULT_BUFFER_SIZE, WriteMode mode = WriteMode::Rewrite) = 0;
+
+    /// Remove file or directory
+    virtual void remove(const String & path, bool recursive) = 0;
+
+public:
+    /// Used for reservation counters modification
+    static std::mutex reservationMutex;
```
Comment (alexey-milovidov) [inline dbms/src/Disks/IDisk.h]: Looks very strange.

File: dbms/src/Disks/DiskS3.cpp [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+            : WriteBufferFromS3(client_ptr_, bucket_, s3_path_, DEFAULT_BLOCK_SIZE, buf_size_)
+            , metadata_path(metadata_path_)
+            , s3_path(s3_path_)
+        {
+        }
+
+        void finalize() override
+        {
+            WriteBufferFromS3::finalize();
+            Poco::FileOutputStream(metadata_path) << s3_path;
```
Comment (alexey-milovidov) [inline dbms/src/Disks/DiskS3.cpp]: Not sure if error is checked.

**Pull request merged**